### PR TITLE
[PDE-4989] feat(schema): Add support for bulk writes

### PR DESCRIPTION
**Pull request by kola-er**

### CODE REVIEW

File: packages/schema/lib/utils/makeValidator.js [hunk 1]
```diff
@@ -130,9 +130,14 @@ const makeValidator = (mainSchema, subSchemas) => {
   return {
     validate: (definition) => {
       const results = v.validate(definition, mainSchema);
-      const allErrors = results.errors.concat(
+      let allErrors = results.errors.concat(
         functionalConstraints.run(definition, mainSchema)
       );
+
+      allErrors = allErrors.filter((error) => {
+        // remove errors with message 'requires property "perform"' when definition[action] has bulk/performBulk set.
```
Comment (kola-er) [inline packages/schema/lib/utils/makeValidator.js:138]: @eliangcs As we still need to make the property `perform` required, I was thinking of removing the errors with the message 'requires property "perform"' when definition[action] has bulk/performBulk set.

File: packages/schema/lib/functional-constraints/bulkWriteConstraints.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+const jsonschema = require('jsonschema');
+
+const bulkWriteConstraints = (definition) => {
+  const errors = [];
+  const actionType = 'creates';
+
+  if (definition[actionType]) {
+    _.each(definition[actionType], (actionDef) => {
+      if (actionDef.operation && actionDef.operation.bulk) {
+        if (!actionDef.operation.performBulk) {
```
Comment (eliangcs) [inline packages/schema/lib/functional-constraints/bulkWriteConstraints.js:13]: The other direction should be checked too: If `performBulk` is defined, `bulk` should be defined too.

**Comment by eliangcs** [comment]
cc @FokkeZB if you'd like to review too.

File: packages/schema/docs/build/schema.md [hunk 3]
```diff
@@ -593,6 +596,34 @@ Key | Required | Type | Description
...
+
+* **Type** - `object`
+* [**Source Code**](https://github.com/zapier/zapier-platform/blob/zapier-platform-schema@15.7.2/packages/schema/lib/schemas/BulkObjectSchema.js)
+
+#### Properties
+
+Key | Required | Type | Description
+--- | -------- | ---- | -----------
+`groupedBy` | **yes** | `array` | The list of keys of input fields to group bulk-write with. The actual user data provided for the fields will be used during execution. Note that a required input field should be referenced to get user data always.
+`limit` | **yes** | `integer` | The maximum number of items to call performBulk with.
```
Comment (FokkeZB) [inline packages/schema/docs/build/schema.md:613]: May be helpful (but also too lengthy here) to explain:

1. How to decide on a good value, based on max payload size.
2. That there's also a time/window limit, which we manage.

File: packages/schema/lib/schemas/BulkObjectSchema.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+module.exports = makeSchema({
+  id: '/BulkObjectSchema',
+  description:
+    'Zapier uses this configuration for writing in bulk.',
+  type: 'object',
+  required: ['groupedBy', 'limit'],
+  properties: {
+    groupedBy: {
+      description:
+        'The list of keys of input fields to group bulk-write with. The actual user data provided for the fields will be used during execution. Note that a required input field should be referenced to get user data always.',
```
[the remark this answers is not in the document]
Comment (FokkeZB) [reply packages/schema/lib/schemas/BulkObjectSchema.js:14]: I'm for starting strict, but I wouldn't be surprised if at some time we'd need to loosen it up to deal with edge cases.

**Review by FokkeZB — APPROVED** [review]
Left a few comments, but otherwise LGTM

File: packages/schema/lib/utils/makeValidator.js (hunk 1 above)
Comment (kola-er) [reply packages/schema/lib/utils/makeValidator.js:138]: Good idea, thanks.

File: packages/schema/lib/functional-constraints/bulkWriteConstraints.js (hunk 2 above)
Comment (kola-er) [reply packages/schema/lib/functional-constraints/bulkWriteConstraints.js:13]: Done

File: packages/schema/lib/schemas/BulkObjectSchema.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kola-er) [reply packages/schema/lib/schemas/BulkObjectSchema.js:14]: Done

File: packages/schema/lib/utils/makeValidator.js (hunk 1 above)
Comment (kola-er) [reply packages/schema/lib/utils/makeValidator.js]: Done

File: packages/schema/lib/functional-constraints/bulkWriteConstraints.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            if (inputField.required) {
+              requiredInputFields.push(inputField.key);
+            }
+          });
+
+          actionDef.operation.bulk.groupedBy.forEach((field, index) => {
+            if (!requiredInputFields.includes(field)) {
+              errors.push(
+                new jsonschema.ValidationError(
+                  `cannot use optional inputField "${field}".`,
```
Comment (eliangcs) [inline packages/schema/lib/functional-constraints/bulkWriteConstraints.js:52]: nitpick:

```suggestion
                  `cannot use optional or non-existent inputField "${field}".`,
```

**Review by eliangcs — APPROVED** [review]
I tested and it works great! I only have one small suggestion on the validation message. Otherwise, I think this is good to merge.

Should we merge this to a temporary branch like `release-15.8.0` until all the bulk action changes is ready?

**Pull request merged**